data-prepper-plugins/csv-processor/src/main/java/com/amazon/dataprepper/plugins/processor/csv/CSVProcessor.java

```diff
@@ -12,22 +12,60 @@
 import com.amazon.dataprepper.model.processor.AbstractProcessor;
 import com.amazon.dataprepper.model.processor.Processor;
 import com.amazon.dataprepper.model.record.Record;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvParser;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+
+import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
 
 /**
  * Processor to parse CSV data in Events.
  *
  */
 @DataPrepperPlugin(name="csv", pluginType = Processor.class, pluginConfigurationType = CSVProcessorConfig.class)
 public class CSVProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private final CSVProcessorConfig config;
+
     @DataPrepperPluginConstructor
     public CSVProcessor(final PluginMetrics pluginMetrics, final CSVProcessorConfig config) {
         super(pluginMetrics);
-        throw new UnsupportedOperationException();
+        this.config = config;
     }
 
     @Override
     public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
+        final CsvMapper mapper = createCsvMapper();
+        final CsvSchema schema = createCsvSchema();
+
+        for (final Record<Event> record : records) {
+
+            final Event event = record.getData();
+
+            final String message = event.get(config.getSource(), String.class);
+            final boolean userDidSpecifyHeaderEventKey = (config.getColumnNamesSourceKey() != null);
+            final boolean thisEventHasHeaderSource = userDidSpecifyHeaderEventKey && event.containsKey(config.getColumnNamesSourceKey());
+
+            try {
+                final MappingIterator<List<String>> messageIterator = mapper.readerFor(List.class).with(schema).readValues(message);
+
+                // otherwise the message is empty
+                if (messageIterator.hasNextValue()) {
```

Comment:
Would it make sense for this to be `while` instead of `if` are we always guaranteed that there would be just one row? How would this work for parsing a whole CSV file at one?

Reply:
This is a good callout and raises more questions about the behavior when there is a multiline newline-delimited CSV file in the `message` field. Right now we assume that splitting the CSV file into a collection of single line records is handled upstream by the codec, so the processor will receive Events with a single row record. With the current implementation, only the first line will be parsed. If we change to `while` instead of `if`, the Event will be overwritten by every line so it will eventually only contain the last line. So neither solution works great for multiline CSV files.

I can think of two ways to handle multiline CSV files off the top of my head:

1. Split each data column into a separate field:
`{"message":"1,2\n3,4"}` => `{"col1":"1,3","col2":"2,4"}`
2. Emit multiple outgoing events. Parse each line of the original `message` field into a separate Event: 
`{"message":"1,2\n3,4"}` => `[{"col1":"1","col2":"2"}, {"col1":"3","col2":"4"}]`
I think that 2 is a better solution because each CSV row record is kept together, which makes the data more interpretable since users can compare the data in different fields on each event.

If we want to support header autodetection in the CSV file in the Event field, we will need to add another configuration option so that users can indicate that the incoming Event will contain a multiline CSV file with a header (and possibly another configuration option for the newline delimiter, which would be `\n` be default). 

What are your thoughts on handling multiline CSV files and adding configuration options for entire CSV file header autodetection?

Reply:
Travis and I discussed processing multiline CSV files offline and decided to table multiline processing for now because the main ways of ingesting CSV files (through the `file` source or `S3 Source`) already split the data into rows. The current processor and config can be extended to handle multiline CSV parsing for a single event in a follow up PR if there's a need for it.